File: armsauv_docking/src/usbl_simulator.cpp
#include <armsauv_docking/common.h>
#include <armsauv_docking/usbl_simulator.h> 

namespace armsauv_docking{
    USBLSimulator::USBLSimulator(tf::TransformListener& tf) : tf_(tf){
        /* Create ros node */
	contentPrint<std::string>("usbl_sim", "Cereate node handle");
	
	ros::NodeHandle nh;
        ros::NodeHandle private_nh("~");

        // tf_ = new tf::TransformListener(ros::Duration(5));

	contentPrint<std::string>("usbl_sim", "Set parameters");

        /* Load parameters */
        private_nh.param("transceiver_frame", transc_frame_, std::string("armsauv/base_link"));
        private_nh.param("input_topic", input_topic_, std::string("/qt3/pose_gt"));
        private_nh.param("output_topic", output_topic_, std::string("/usbl/output"));
	// private_nh.param("period", period_, 0.05);

        /* Create components */
        usbl_pub_ = nh.advertise<armsauv_msgs::Usbl>(output_topic_, 1);
        transp_sub_ = nh.subscribe<nav_msgs::Odometry>(input_topic_, 1, boost::bind(&USBLSimulator::transpCb, this, _1));

	contentPrint<std::string>("usbl_sim", "Start thread");

        /* Start publish thread */
        pub_thread_ = new boost::thread(boost::bind(&USBLSimulator::outputPublish, this)); 
    }

    USBLSimulator::~USBLSimulator(){
        if(pub_thread_){
            delete pub_thread_;
            pub_thread_ = nullptr;
        }

	// if(tf_){
	//     delete tf_;
	//     tf_ = nullptr;
	// }
    }

    void USBLSimulator::transpCb(const nav_msgs::Odometry::ConstPtr& input){
        /* Process input */
	/*
        geometry_msgs::PoseStamped transp_to_world;
        std::string transp_frame = input->header.frame_id;
        transp_to_world.header.frame_id = transp_frame; 
        transp_to_world.header.stamp = input->header.stamp;
        transp_to_world.pose.position.x = input->pose.pose.position.x;
        transp_to_world.pose.position.y = input->pose.pose.position.y;
        transp_to_world.pose.position.z = input->pose.pose.position.z;
        transp_to_world.pose.orientation.x = input->pose.pose.orientation.x;
        transp_to_world.pose.orientation.y = input->pose.pose.orientation.y;
        transp_to_world.pose.orientation.z = input->pose.pose.orientation.z;
        transp_to_world.pose.orientation.w = input->pose.pose.orientation.w;
	*/
        geometry_msgs::PointStamped transp_to_world;
        std::string transp_frame = input->header.frame_id;
        transp_to_world.header.frame_id = transp_frame; 
        transp_to_world.header.stamp = input->header.stamp;
        transp_to_world.point.x = input->pose.pose.position.x;
        transp_to_world.point.y = input->pose.pose.position.y;
        transp_to_world.point.z = input->pose.pose.position.z;
        
        // geometry_msgs::PoseStamped transp_to_transc;
	geometry_msgs::PointStamped transp_to_transc; 

        /* Transform transponder pose in world frame to pose in transceiver frame */
        try{
            // tf_.transformPose(transc_frame_, ros::Time(0), transp_to_world, transp_frame, transp_to_transc);
	    tf_.transformPoint(transc_frame_, ros::Time(0), transp_to_world, transp_frame, transp_to_transc); 
	}
        catch(tf::TransformException& ex){
            ROS_ERROR("Received exception trying ot transform transponder pose to transceiver frame: %s", ex.what());
            return ;
        }

        /* Update transceiver message */
	std::cout << "[usbl_sim]:" << "local_frame:" << transp_frame << " target_frame:" << transp_to_transc.header.frame_id << std::endl;

        /*
        double d_x = transp_to_transc.pose.position.x;
        double d_y = transp_to_transc.pose.position.y;
        double d_z = transp_to_transc.pose.position.z;
        */

        double d_x = transp_to_transc.point.x;
        double d_y = transp_to_transc.point.y;
        double d_z = transp_to_transc.point.z;
	
        std::cout << "[usbl_sim]:" << "delta x:" << d_x
		                   << " delta y:" << d_y
				   << " delta z:" << d_z << std::endl;

        /*
	tf::Quaternion temp_q;
        tf::quaternionMsgToTF(transp_to_transc.pose.orientation, temp_q);

        tf::Matrix3x3 temp_m(temp_q);
        */

        double roll, pitch, yaw;
        /*
        temp_m.getRPY(roll, pitch, yaw);
	pitch = -pitch; yaw = -yaw;
        */
        roll = atan2(d_z,d_y); 
        pitch = atan2(d_z,d_x);
        yaw = atan2(d_y,d_x);

	std::cout << "[usbl_sim]:" << "roll:" << roll
		                   << " pitch:" << pitch
				   << " yaw:" << yaw << std::endl;

        std::lock_guard<std::mutex> lckg(transc_mutex_);
        transc_msg_.phi = yaw;
        transc_msg_.s_r = std::sqrt(d_x*d_x + d_y*d_y + d_z*d_z);
        transc_msg_.h_r = std::sqrt(d_x*d_x + d_y*d_y);
        transc_msg_.psi = std::atan(d_z / transc_msg_.h_r); 
    }

    void USBLSimulator::outputPublish(){
        ros::NodeHandle nh;
        while(nh.ok()){
            std::lock_guard<std::mutex> lckg(transc_mutex_);
            usbl_pub_.publish(transc_msg_); // pub

            boost::this_thread::sleep(boost::posix_time::milliseconds(period_ * 1000));
        }
    }
}; // ns

File: armsauv_docking/scripts/usbl_simulator.py
#!/usr/bin/env python

import math
import rospy
import sys
import tf
from tf import TransformListener
from nav_msgs.msg import Odometry
from std_msgs.msg import Float64
from geometry_msgs.msg import PoseStamped
from armsauv_msgs.msg import Usbl

class USBLSimulator:
    def __init__(self):
        ''' Simulate USBL message according to global position '''
        # get parameters
        self.getConfig()

        # components initialization
        rospy.init_node("usbl_simulator")

        self.tf_tree_ = TransformListener()

        rospy.Subscriber(self.input_topic_, Odometry, self.getInput)
        self.input_ = Odometry()

        self.output_pub_ = rospy.Publisher(self.output_topic_, Usbl, queue_size=1)
        self.output_ = Usbl()

    def getConfig(self):
        # get frame
        self.transc_frame_ = rospy.get_param('transceiver frame', '/armsauv/base_link')

        # get period
        self.period_ = rospy.get_param('period', 0.1)

        # get input topic
        self.input_topic_ = rospy.get_param('input topic', '/qt4/pose_gt')

        # get output topic
        self.output_topic_ = rospy.get_param('output topic', '/usbl/output')

    def getInput(self, input_pose):
        # Update input
        self.input_ = input_pose

    def getParams(self):
        # Transform pose of transponder from global frame to frame of transceiver
        input_to_transc = PoseStamped()

        input_to_global = PoseStamped()
        input_to_global.header.frame_id = self.input_.header.frame_id
        # print(self.input_.header.frame_id)
        input_to_global.header.stamp = self.input_.header.stamp
        input_to_global.pose.position = self.input_.pose.pose.position
        input_to_global.pose.orientation= self.input_.pose.pose.orientation

        self.tf_tree_.waitForTransform(self.transc_frame_, input_to_global.header.frame_id, rospy.Time.now(), rospy.Duration(0.5))
        while not rospy.is_shutdown():
            try:
                self.tf_tree_.waitForTransform(self.transc_frame_, input_to_global.header.frame_id, rospy.Time.now(), rospy.Duration(0.5))
                input_to_transc = self.tf_tree_.transformPose(self.transc_frame_, input_to_global)
            except rospy.ROSInterruptException: pass
        
        '''
        try:
            input_to_transc = self.tf_tree_.transformPose(self.transc_frame_, input_to_global)
        except rospy.ROSInterruptException: 
            pass
        '''
        input_to_transc_o = input_to_transc.pose.orientation
        input_to_transc_l = [input_to_transc_o.x, input_to_transc_o.y, input_to_transc_o.z, input_to_transc_o.w]
        euler_angle = tf.transformations.euler_from_quaternion(input_to_transc_l)
        
        x = input_to_transc.pose.position.x
        y = input_to_transc.pose.position.y
        z = input_to_transc.pose.position.z

        # get yaw to transceiver frame
        phi = euler_angle[2]

        # get euclidean distance
        dist = sqrt(x*x + y*y + z*z)

        # get horizontal plane project distance
        h_dist = sqrt(x*x + y*y)

        # get lateral angle difference
        psi = atan(z / h_dist)

        # print
        print("USBL: ", phi, psi, h_dist, dist)

        return [phi, psi, h_dist, dist]

    def outputPublish(self):
        # publish usbl messages
        output = self.getParams()
        self.output_.phi = output[0]
        self.output_.psi = output[1]
        self.output_.h_r = output[2]
        self.output_.s_r = output[3]

        self.output_pub_.publish(self.output)

        rospy.sleep(self.period_)
            
if __name__ == '__main__':
    try:
        usbl_simulator = USBLSimulator()
        rospy.sleep(3.0)

        while not rospy.is_shutdown():
            usbl_simulator.outputPublish()
    except rospy.ROSInterruptException: pass
